core/src/main/java/io/grpc/internal/BinaryLogProvider.java

```diff
@@ -141,15 +144,72 @@ public void close() throws IOException {
   protected abstract boolean isAvailable();
 
   // Creating a named class makes debugging easier
-  private static final class IdentityMarshaller implements Marshaller<InputStream> {
+  private static final class ByteArrayMarshaller implements Marshaller<byte[]> {
+    @Override
+    public InputStream stream(byte[] value) {
+      return new ByteArrayRetainedInputStream(value);
+    }
+
+    @Override
+    public byte[] parse(InputStream stream) {
+      if (stream instanceof ByteArrayRetainedInputStream) {
```

Comment:
This optimization seems unnecessary if you cast `T` directly to `byte[]`, as you won't need to do an unnecessary `stream` /`parse` roundtrip (you just return T as you received it).